packages/realm/binding/android/src/main/cpp/io_realm_react_RealmReactPackage.cpp

```diff
@@ -0,0 +1,150 @@
+
+    __android_log_print(ANDROID_LOG_VERBOSE, "Realm", "Injecting module into JS global");
+    auto runtime = reinterpret_cast<jsi::Runtime*>(runtime_pointer);
+    if (!runtime) {
+        return;
+    }
+
+    auto exports = jsi::Object(*runtime);
+    realm_jsi_init(*runtime, exports);
+    runtime->global().setProperty(*runtime, "__injectedRealmBinding", exports);
+}
+
+extern "C"
+JNIEXPORT void JNICALL
+Java_io_realm_react_RealmReactPackage_injectCallInvoker(JNIEnv *env, jclass clazz,
+                                                         jobject call_invoker) {
+    __android_log_print(ANDROID_LOG_VERBOSE, "Realm", "Getting JS call invoker");
+    // React Native uses the fbjni library for handling JNI, which has the concept of "hybrid objects",
+    // which are Java objects containing a pointer to a C++ object. The CallInvokerHolder, which has the
+    // invokeAsync method we want access to, is one such hybrid object.
+    // Rather than reworking our code to use fbjni throughout, this code unpacks the C++ object from the Java
+    // object `callInvokerHolderJavaObj` manually, based on reverse engineering the fbjni code.
+
+    // 1. Get the Java object referred to by the mHybridData field of the Java holder object
+    auto callInvokerHolderClass = env->GetObjectClass(call_invoker);
+    auto hybridDataField = env->GetFieldID(callInvokerHolderClass, "mHybridData", "Lcom/facebook/jni/HybridData;");
+    auto hybridDataObj = env->GetObjectField(call_invoker, hybridDataField);
+
+    // 2. Get the destructor Java object referred to by the mDestructor field from the myHybridData Java object
+    auto hybridDataClass = env->FindClass("com/facebook/jni/HybridData");
+    auto destructorField =
+            env->GetFieldID(hybridDataClass, "mDestructor", "Lcom/facebook/jni/HybridData$Destructor;");
+    auto destructorObj = env->GetObjectField(hybridDataObj, destructorField);
+
+    // 3. Get the mNativePointer field from the mDestructor Java object
+    auto destructorClass = env->FindClass("com/facebook/jni/HybridData$Destructor");
+    auto nativePointerField = env->GetFieldID(destructorClass, "mNativePointer", "J");
+    auto nativePointerValue = env->GetLongField(destructorObj, nativePointerField);
+
+    // 4. Cast the mNativePointer back to its C++ type
+    auto nativePointer = reinterpret_cast<facebook::react::CallInvokerHolder*>(nativePointerValue);
+
+    // 5. Store the JS call invoker for the workaround to use
+    realm::js::flush_ui_workaround::inject_js_call_invoker(nativePointer->getCallInvoker());
+}
+
+extern "C"
+JNIEXPORT void JNICALL
+Java_io_realm_react_RealmReactPackage_invalidateCaches(JNIEnv *env, jclass clazz) {
+    // Disable the flush ui workaround
+    realm::js::flush_ui_workaround::reset_js_call_invoker();
+    __android_log_print(ANDROID_LOG_VERBOSE, "Realm", "Invalidating caches");
+#if DEBUG
+    realm_jsi_close_sync_sessions();
+#endif
+    realm_jsi_invalidate_caches();
+}
+
+#endif
```

Comment:
I've moved all the JNI code from the `RealmReactModule` to static native members of the `RealmReactPackage` (since their implementations weren't tied to the lifetime of the `RealmReactModule` anyway).